### PR TITLE
resolves #1516 add option to exclude tags when including file

### LINKED ISSUE #1516: Allow the option to exclude tags when using the include directive
Sometimes you want to include parts of a document, but exclude other parts (child tags). I'd like to see something like this:

include::presentation.asc[tag=content,exclude-tag=demo1]

or:

include::presentation.asc[tag=content,exclude-tags=demo1;demo4]

### PR DESCRIPTION
**Pull request by mojavelinux**
- in addition to including tags, add option to exclude tags using !name syntax
- introduce wildcard for including all (\*) or excluding all (!*) tags
- introduce wildcard for selecting non-tagged lines (**)
- don't select nested tags with wildcard if ancestor tag is excluded
- warn if a mismatched or unexpected end tag is found
- improve efficiency of detecting and processing tag directives
- maintain a tag stack for handling nested tags
- skip call to push_include if selected lines array is empty
- ignore tag and tags attribute if empty (previously caused an infinite loop)
- add tests for excluding tags

### CODE REVIEW

**Comment by mojavelinux** [comment]
I'm still missing tests to verify that a warning is issued if an unexpected or misplaced end tag is found.

File: lib/asciidoctor/substitutors.rb [hunk 1]
```diff
@@ -728,19 +728,23 @@ def sub_macros(source)
       result = result.gsub(LinkInlineRx) {
         # alias match for Ruby 1.8.7 compat
         m = $~
+        # NOTE if text is non-nil, this the macro syntax
```
Comment (Mogztter) [inline lib/asciidoctor/substitutors.rb]: I don't quite get this comment "this the macro syntax" ?

File: lib/asciidoctor/substitutors.rb [hunk 2]
```diff
@@ -728,19 +728,23 @@ def sub_macros(source)
...
+        text = m[3]
         # honor the escape
         if m[2].start_with? RS
-          next %(#{m[1]}#{m[2][1..-1]}#{m[3]})
+          next %(#{m[1]}#{m[2][1..-1]}#{text})
```
Comment (Mogztter) [inline lib/asciidoctor/substitutors.rb]: I think we should aldo named `m[0]`, `m[1]` and `m[2]`

File: lib/asciidoctor/substitutors.rb (hunk 1 above)
Comment (mojavelinux) [inline lib/asciidoctor/substitutors.rb]: The changes to substitutors.rb weren't supposed to be part of this PR. I'm going to pull them out into a separate PR because they are addressing something totally different.

**Comment by mojavelinux** [comment]
This PR is now ready for review.
